src/clds_hash_table.c

```diff
@@ -929,6 +1029,247 @@ CLDS_HASH_TABLE_ITEM* clds_hash_table_find(CLDS_HASH_TABLE_HANDLE clds_hash_tabl
+        BUCKET_ARRAY* current_bucket_array = (BUCKET_ARRAY*)InterlockedCompareExchangePointer((volatile PVOID*)&clds_hash_table->first_hash_table, NULL, NULL);
+        while (current_bucket_array != NULL)
+        {
+            BUCKET_ARRAY* next_bucket_array = (BUCKET_ARRAY*)InterlockedCompareExchangePointer((volatile PVOID*)&current_bucket_array->next_bucket, NULL, NULL);
+
+            if (InterlockedAdd(&current_bucket_array->item_count, 0) != 0)
+            {
+                LONG bucket_count = InterlockedAdd(&current_bucket_array->bucket_count, 0);
+                LONG i;
+                for (i = 0; i < bucket_count; i++)
+                {
+                    if (current_bucket_array->hash_table[i] != NULL)
+                    {
+                        /* Codes_SRS_CLDS_HASH_TABLE_42_020: [ clds_hash_table_snapshot shall call clds_sorted_list_lock_writes. ]*/
+                        clds_sorted_list_lock_writes(current_bucket_array->hash_table[i]);
+                        need_to_unlock_all = true;
+
+                        /* Codes_SRS_CLDS_HASH_TABLE_42_021: [ clds_hash_table_snapshot shall call clds_sorted_list_get_count and add to the running total. ]*/
+                        uint64_t list_item_count;
+                        CLDS_SORTED_LIST_GET_COUNT_RESULT count_result = clds_sorted_list_get_count(current_bucket_array->hash_table[i], clds_hazard_pointers_thread, &list_item_count);
+                        if (count_result != CLDS_SORTED_LIST_GET_COUNT_OK)
+                        {
+                            /* Codes_SRS_CLDS_HASH_TABLE_42_061: [ If there are any other failures then clds_hash_table_snapshot shall fail and return CLDS_HASH_TABLE_SNAPSHOT_ERROR. ]*/
+                            LogError("clds_sorted_list_get_count failed with %" PRI_MU_ENUM, MU_ENUM_VALUE(CLDS_SORTED_LIST_GET_COUNT_RESULT, count_result));
+                            clds_sorted_list_unlock_writes(current_bucket_array->hash_table[i]);
+                            break;
+                        }
+                        else
+                        {
+                            if (list_item_count + temp_item_count < temp_item_count)
+                            {
+                                /* Codes_SRS_CLDS_HASH_TABLE_42_022: [ If the addition of the list count causes overflow then clds_hash_table_snapshot shall fail and return CLDS_HASH_TABLE_SNAPSHOT_ERROR. ]*/
+                                LogError("overflow in computing total count (%" PRIu64 " + %" PRIu64 ")", temp_item_count, list_item_count);
+                                clds_sorted_list_unlock_writes(current_bucket_array->hash_table[i]);
+                                break;
+                            }
+                            else
+                            {
+                                temp_item_count += list_item_count;
+                            }
+                        }
+                    }
+                }
```

Comment:
I think we should just be using item_count here for each bucket array? Since we only start snapshotting when all writes are done, this should be accurate. Was there a reason to not do that?

Reply:
I think this is correct (but I was sort of "reverse engineering" this to figure out how it works and you actually wrote it).

From what I see, the `hash_table` is index modulo the bucket_count:


```
            bucket_index = hash % InterlockedAdd(&find_bucket_array->bucket_count, 0);
            bucket_list = InterlockedCompareExchangePointer(&find_bucket_array->hash_table[bucket_index], NULL, NULL);
```

So I iterate over everything from 0 to `bucket_count`. And I just check that `item_count` is non-zero so I can skip over the empty buckets.

---
In reply to: [352379325](https://github.com/dcristoloveanu/clds/pull/33#discussion_r352379325) [](ancestors = 352379325)